### ISSUE TITLE
pkg --process failing in pipenv

### ISSUE THREAD

**Issue by douglasmccormickjr** [Digital-Sapphire/PyUpdater#143]
I think I'm very close to getting things to work correctly, I'm getting tripped up on one little step.  Let me get to the issue....

I believe I need to run the "pyupdater pkg -s -p" step in order to both process and sign the application that was recently built.  The "--sign" step seems to run fine alone (albeit it just creates the 2 files "keys.gz" and "versions.gz" and puts both files into the "deploy folder.  I should mention though, when running the "--sign" step alone, the newly built application (located in the "new" folder) doesn't get moved into the deploy folder-- it doesn't look like it gets touched at all.  This leads me to believe I need to run the "--process" step to have the application correctly get signed and ready for scp deployment. Please correct me if my assumptions are wrong.

Here's what happens when I run that step involving "--process"....

    C:\PythonEXE\PyyUp>pipenv run pyupdater pkg -s -p
    355 INFO: PyUpdater 2.5.3
    [INFO] PyUpdater 2.5.3
    360 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    360 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    360 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    361 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    361 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    361 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    362 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    362 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    363 INFO: Patch support enabled
    [INFO] Patch support enabled
    363 INFO: Creating new config file
    [INFO] Creating new config file
    369 DEBUG: found extension EntryPoint.parse('scp = scp_uploader:SCPUploader')
    457 INFO: Processing packages...
    [INFO] Processing packages...
    459 INFO: Getting package list
    [INFO] Getting package list
    459 DEBUG: Changing to Directory --> C:\PythonEXE\PyyUp\pyu-data\new
    459 DEBUG: Extracting update archive info for: DougsApp-win-1.11.0.zip
    461 DEBUG: Version str: DougsApp-win-1.11.0.zip
    461 DEBUG: Removed ".zip"
    461 DEBUG: Got version info
    461 DEBUG: Parsing "DougsApp-win-1.11.0.zip" for platform info
    461 DEBUG: Platform name is: win
    461 DEBUG: Got platform info
    461 DEBUG: Package name: DougsApp-win-1.11.0.zip
    461 DEBUG: Regex name: DougsApp
    462 DEBUG: Got name of update: DougsApp
    462 DEBUG: Info extraction complete
    462 DEBUG: Getting package hashes
    487 DEBUG: Hash for file C:\PythonEXE\PyyUp\pyu-data\new\DougsApp-win-1.11.0.zip: fff677ab4bc0349e6c1fbdfa922ea1379893a53052fe71df9a61e019125917f8
    488 DEBUG: File size: 7451905 bytes
    488 DEBUG: Moving back to Directory --> C:\PythonEXE\PyyUp
    'updates'
    489 ERROR: 'updates'
    [ERROR] 'updates'
    489 DEBUG: 'updates'
    Traceback (most recent call last):
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\__init__.py", line 119, in main
        _real_main(args)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\__init__.py", line 88, in _real_main
        dispatch_command(args, pyi_args)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\__init__.py", line 105, in dispatch_command
        cmd(args, pyi_args)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\commands.py", line 297, in _cmd_pkg
        pyu.process_packages(ns.verbose)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\core.py", line 77, in process_packages
        self.ph.process_packages(report_errors)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\package_handler\__init__.py", line 116, in process_packages
        pkg_manifest, patch_manifest = self._get_package_list(report_errors)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\package_handler\__init__.py", line 198, in _get_package_list
        package)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\package_handler\__init__.py", line 336, in _update_file_list
        files = json_data[settings.UPDATES_KEY]
    KeyError: 'updates'

    C:\PythonEXE\PyyUp>

My Environment:
Windows 10
PyUpdater 2.5.3
Python 3.6.5

I went ahead and tried to debug/step through the variables that caused the "KeyError: 'updates'" thrown error.  From what I've uncovered, it looks like the "settings.UPDATES_KEY" is getting pulled from the file:
C:\Users\domccorm\.virtualenvs\PyyUp-6sVZMsOM\Lib\site-packages\pyupdater\settings.py

I hope that helps a bit.  Any advice would be greatly appreciated.

Thanks,
-Doug Jr.

**Comment by 2press** [comment]
Have you tried running `pyupdater pkg --process` first and then `pyupdater pkg --sign`? The solution might just be to swap your arguments and use `pyupdater pkg --process --sign` as you have to process before signing. 

Further questions: Have you completed the `pyupdater init`, keys and `pyupdater build` process?

**Comment by douglasmccormickjr** [comment]
I tried the "-p" flag and the  "-p -s" (swapping the order) --- I'm still getting the same errors.

Here's the dump from the console:

     C:\PythonEXE\PyyUp>pipenv run pyupdater pkg -p
    1444 INFO: PyUpdater 2.5.3
    [INFO] PyUpdater 2.5.3
    1450 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    1450 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    1455 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    1457 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    1458 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    1458 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    1459 DEBUG: Config Dir: C:\PythonEXE\PyyUp\.pyupdater
    1459 DEBUG: Config DB: C:\PythonEXE\PyyUp\.pyupdater\config.pyu
    1460 INFO: Patch support enabled
    [INFO] Patch support enabled
    1461 INFO: Creating new config file
    [INFO] Creating new config file
    1517 DEBUG: found extension EntryPoint.parse('scp = scp_uploader:SCPUploader')
    2408 INFO: Processing packages...
    [INFO] Processing packages...
    2410 INFO: Getting package list
    [INFO] Getting package list
    2410 DEBUG: Changing to Directory --> C:\PythonEXE\PyyUp\pyu-data\new
    2410 DEBUG: Extracting update archive info for: DougsApp-win-1.11.0.zip
    2411 DEBUG: Version str: DougsApp-win-1.11.0.zip
    2411 DEBUG: Removed ".zip"
    2412 DEBUG: Got version info
    2412 DEBUG: Parsing "DougsApp-win-1.11.0.zip" for platform info
    2412 DEBUG: Platform name is: win
    2412 DEBUG: Got platform info
    2412 DEBUG: Package name: DougsApp-win-1.11.0.zip
    2412 DEBUG: Regex name: DougsApp
    2413 DEBUG: Got name of update: DougsApp
    2413 DEBUG: Info extraction complete
    2413 DEBUG: Getting package hashes
    2458 DEBUG: Hash for file C:\PythonEXE\PyyUp\pyu-data\new\DougsApp-win-1.11.0.zip: fff677ab4bc0349e6c1fbdfa922ea1379893a53052fe71df9a61e019125917f8
    2460 DEBUG: File size: 7451905 bytes
    2460 DEBUG: Moving back to Directory --> C:\PythonEXE\PyyUp
    'updates'
    2461 ERROR: 'updates'
    [ERROR] 'updates'
    2461 DEBUG: 'updates'
    Traceback (most recent call last):
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\__init__.py", line 119, in main
        _real_main(args)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\__init__.py", line 88, in _real_main
        dispatch_command(args, pyi_args)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\__init__.py", line 105, in dispatch_command
        cmd(args, pyi_args)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\cli\commands.py", line 297, in _cmd_pkg
        pyu.process_packages(ns.verbose)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\core.py", line 77, in process_packages
        self.ph.process_packages(report_errors)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\package_handler\__init__.py", line 116, in process_packages
        pkg_manifest, patch_manifest = self._get_package_list(report_errors)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\package_handler\__init__.py", line 198, in _get_package_list
        package)
      File "c:\users\domccorm\.virtualenvs\pyyup-6svzmsom\lib\site-packages\pyupdater\package_handler\__init__.py", line 336, in _update_file_list
        files = json_data[settings.UPDATES_KEY]
    KeyError: 'updates'

    C:\PythonEXE\PyyUp>


As for the "init" and "build" -- those look to be working well (app compiles and zips correctly)...I also had to do the little pure25519 trick you were working on (it did create and import the keys correctly as far as I can tell) -- thank you for coming up with that change, it helped a ton...I should mention, I think pipenv is a bit neurotic about versioning-- here's what happens when you install it (it doesn't worry me too much though)

    C:\PythonEXE\PyyUp>pipenv install pure25519
    Installing pure25519…
    Adding pure25519 to Pipfile's [packages]…
    Installation Succeeded
    Pipfile.lock (d40d69) out of date, updating to (4ce809)…
    Locking [dev-packages] dependencies…
    Locking [packages] dependencies…
    Locking Failed!
    [pipenv.exceptions.ResolutionFailure]:   File "c:/python360564/lib/site-packages/pipenv/resolver.py", line 69, in resolve
    [pipenv.exceptions.ResolutionFailure]:       req_dir=requirements_dir
    [pipenv.exceptions.ResolutionFailure]:   File "c:\python360564\lib\site-packages\pipenv\utils.py", line 726, in resolve_deps
    [pipenv.exceptions.ResolutionFailure]:       req_dir=req_dir,
    [pipenv.exceptions.ResolutionFailure]:   File "c:\python360564\lib\site-packages\pipenv\utils.py", line 480, in actually_resolve_deps
    [pipenv.exceptions.ResolutionFailure]:       resolved_tree = resolver.resolve()
    [pipenv.exceptions.ResolutionFailure]:   File "c:\python360564\lib\site-packages\pipenv\utils.py", line 395, in resolve
    [pipenv.exceptions.ResolutionFailure]:       raise ResolutionFailure(message=str(e))
    [pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches 2to3
    [pipenv.exceptions.ResolutionFailure]:       No versions found
    [pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
      First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.
     Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
      Hint: try $ pipenv lock --pre if it is a pre-release dependency.
    ERROR: ERROR: Could not find a version that matches 2to3
    No versions found
    Was https://pypi.org/simple reachable?
    [pipenv.exceptions.ResolutionFailure]:       req_dir=requirements_dir
    [pipenv.exceptions.ResolutionFailure]:   File "c:\python360564\lib\site-packages\pipenv\utils.py", line 726, in resolve_deps
    [pipenv.exceptions.ResolutionFailure]:       req_dir=req_dir,
    [pipenv.exceptions.ResolutionFailure]:   File "c:\python360564\lib\site-packages\pipenv\utils.py", line 480, in actually_resolve_deps
    [pipenv.exceptions.ResolutionFailure]:       resolved_tree = resolver.resolve()
    [pipenv.exceptions.ResolutionFailure]:   File "c:\python360564\lib\site-packages\pipenv\utils.py", line 395, in resolve
    [pipenv.exceptions.ResolutionFailure]:       raise ResolutionFailure(message=str(e))
    [pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches 2to3
    [pipenv.exceptions.ResolutionFailure]:       No versions found
    [pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
      First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.
     Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
      Hint: try $ pipenv lock --pre if it is a pre-release dependency.
    ERROR: ERROR: Could not find a version that matches 2to3
    No versions found
    Was https://pypi.org/simple reachable?



Back to the first issue at hand with the "--process" step...Perhaps this is somehow connected to pipenv?  I'm going to try a few more things and post them back to this thread.

Any suggestions or help is greatly appreciated.

**Comment by 2press** [comment]
Have you checked your `config.pyu` for corruption (did you make any manual changes to it) - is there a 'updates' key?

**Comment by douglasmccormickjr** [comment]
I tested pyupdater with a "non" pipenv install (I sometimes skip the whole virtual environment thing altogether) and it seems to be working great....I can accomplish all steps (start to finish) and the app is downloading/updating nicely.

I did need to do the edit for "ed25519" with is one liner on the 2 key files that point to that module:

    #import ed25519
    from pure25519 import ed25519_oop as ed25519

The 2 files (for the curious in the future -- this might be patched though):
- c:\python360564\lib\site-packages\pyupdater\key_handler\keys.py
- c:\python360564\lib\site-packages\pyupdater\key_handler\__init__.py

Checking the old install (using pipenv) there doesn't seem to be an "updates" parent/branch in the config.pyu JSON tree...I've checked on my non-pipenv install and that branch does exist-- I think you've found where it's breaking.

I'll do a fresh/new pipenv install to see if enabling "patch updates" is on (that was something I did initially skip initially) -- perhaps that's messing things up.  Also, since pipenv makes up these other directories for modules, maybe on build, the system isn't finding everything correctly?  Not sure really.

**Comment by SIVLEOL** [comment]
I was getting a very similar (KeyError: 'updates') problem and (maybe) resolved it. (I'm using venv through pip-Win)

I first reset the PyUpdater stuff by: 
`pyupdater clean`

I then went through the usual steps:
`pyupdater keys -c`
`py updater init`
`pyupdater keys --import`

Then, I did something differently. Previously I skipped the step of integrating PyInstaller (see: [official documentation](http://www.pyupdater.org/usage-client/)) into the application. (I was following [this demo](https://pyupdater-wx-demo.readthedocs.io/en/latest/overview.html) which didn't show that step because it had already been done, maybe this is the cause of the issue?)

Additionally, previously I had used "pyupdater make-spec ..." (since I wanted pyinstaller to move some folders in for me, I needed to modify win.spec a bit) and then built from that. This time I tried just building directly. ("pyupdater build ...") So my additional folders are missing, but this time running "pyupdater pkg --process --sign" worked without problem. (and I could confirm that "updates" is in config.pyu)

Just putting this out here in case someone else runs into the same error. I'm going to try building via "make-spec" again now.
